assertion failed: asTerm called on not-a-Term val <none>

## Minimized code

The following code compiles on scala 2.x, but fails on dotty 0.23-RC1 and latest nightly. 

https://github.com/DieBauer/dotty-java/blob/master/src/main/scala/Jersey.scala#L17-L20

## Output (click arrow to expand)
<details>

```

[info] Compiling 1 Scala source to /Users/jens/Developer/spring/target/scala-0.24/classes ...
exception while typing MyFactoryProvider.this of class class dotty.tools.dotc.ast.Trees$This # 27886
exception while typing MyFactoryProvider.this.mpep of class class dotty.tools.dotc.ast.Trees$Select # 28110
exception while typing override def get(): 
  
    org.glassfish.jersey.server.internal.inject.
      MultivaluedParameterExtractorProvider
  
 = MyFactoryProvider.this.mpep of class class dotty.tools.dotc.ast.Trees$DefDef # 28111
exception while typing final class $anon() extends Object(), 
  javax.inject.Provider[
    org.glassfish.jersey.server.internal.inject.
      MultivaluedParameterExtractorProvider
  ]
 {| => root / Compile / compileIncremental 0s
  override def get(): 
    
      org.glassfish.jersey.server.internal.inject.
        MultivaluedParameterExtractorProvider
    
   = MyFactoryProvider.this.mpep
} of class class dotty.tools.dotc.ast.Trees$TypeDef # 28113
exception while typing {
  final class $anon() extends Object(), 
    javax.inject.Provider[
      org.glassfish.jersey.server.internal.inject.
        MultivaluedParameterExtractorProvider
    ]
   {
    override def get(): 
      
        org.glassfish.jersey.server.internal.inject.
          MultivaluedParameterExtractorProvider
      
     = MyFactoryProvider.this.mpep
  }
  new Object with javax.inject.Provider {...}():
    
      javax.inject.Provider[
        org.glassfish.jersey.server.internal.inject.
          MultivaluedParameterExtractorProvider
      ]
    
} of class class dotty.tools.dotc.ast.Trees$Block # 28114
exception while typing private <static> def MyFactoryProvider$superArg$1(
  mpep: 
    
      org.glassfish.jersey.server.internal.inject.
        MultivaluedParameterExtractorProvider
    
, locator: org.glassfish.hk2.api.ServiceLocator): 
  
    javax.inject.Provider[
      org.glassfish.jersey.server.internal.inject.
        MultivaluedParameterExtractorProvider
    ]
  
 = 
  {
    final class $anon() extends Object(), 
      javax.inject.Provider[
        org.glassfish.jersey.server.internal.inject.
          MultivaluedParameterExtractorProvider
      ]
     {
      override def get(): 
        
          org.glassfish.jersey.server.internal.inject.
            MultivaluedParameterExtractorProvider
        
       = MyFactoryProvider.this.mpep
    }
    new Object with javax.inject.Provider {...}():
      
        javax.inject.Provider[
          org.glassfish.jersey.server.internal.inject.
            MultivaluedParameterExtractorProvider
        ]
      
  } of class class dotty.tools.dotc.ast.Trees$DefDef # 28120
exception while typing @scala.annotation.internal.SourceFile(
  "/Users/jens/Developer/spring/src/main/scala/Main.scala"
) class MyFactoryProvider @javax.inject.Inject()(
  mpep: 
    
      org.glassfish.jersey.server.internal.inject.
        MultivaluedParameterExtractorProvider
    
, locator: org.glassfish.hk2.api.ServiceLocator) extends 
  org.glassfish.jersey.server.internal.inject.AbstractValueParamProvider
(MyFactoryProvider#MyFactoryProvider$superArg$1(this.mpep, this.locator), 
  [UNKNOWN : org.glassfish.jersey.model.Parameter.Source]
) {
  private val mpep: 
    
      org.glassfish.jersey.server.internal.inject.
        MultivaluedParameterExtractorProvider
    
  private val locator: org.glassfish.hk2.api.ServiceLocator
  override def createValueProvider(
    parameter: org.glassfish.jersey.server.model.Parameter
  ): 
    java.util.function.Function[org.glassfish.jersey.server.ContainerRequest, ?]
   = 
    {
      def $anonfun(request: org.glassfish.jersey.server.ContainerRequest): Any
         = 
      ""
      closure($anonfun:
        
          java.util.function.Function[
            org.glassfish.jersey.server.ContainerRequest
          , ?]
        
      )
    }
  private <static> def MyFactoryProvider$superArg$1(
    mpep: 
      
        org.glassfish.jersey.server.internal.inject.
          MultivaluedParameterExtractorProvider
      
  , locator: org.glassfish.hk2.api.ServiceLocator): 
    
      javax.inject.Provider[
        org.glassfish.jersey.server.internal.inject.
          MultivaluedParameterExtractorProvider
      ]
    
   = 
    {
      final class $anon() extends Object(), 
        javax.inject.Provider[
          org.glassfish.jersey.server.internal.inject.
            MultivaluedParameterExtractorProvider
        ]
       {
        override def get(): 
          
            org.glassfish.jersey.server.internal.inject.
              MultivaluedParameterExtractorProvider
          
         = MyFactoryProvider.this.mpep
      }
      new Object with javax.inject.Provider {...}():
        
          javax.inject.Provider[
            org.glassfish.jersey.server.internal.inject.
              MultivaluedParameterExtractorProvider
          ]
        
    }
} of class class dotty.tools.dotc.ast.Trees$TypeDef # 28131
exception while typing package <empty> {
  @scala.annotation.internal.SourceFile(
    "/Users/jens/Developer/spring/src/main/scala/Main.scala"
  ) class Handler() extends 
    
      org.springframework.web.servlet.mvc.method.annotation.
        ResponseEntityExceptionHandler
    
  () {}
  @scala.annotation.internal.SourceFile(
    "/Users/jens/Developer/spring/src/main/scala/Main.scala"
  ) class MyFactoryProvider @javax.inject.Inject()(
    mpep: 
      
        org.glassfish.jersey.server.internal.inject.
          MultivaluedParameterExtractorProvider
      
  , locator: org.glassfish.hk2.api.ServiceLocator) extends 
    org.glassfish.jersey.server.internal.inject.AbstractValueParamProvider
  (MyFactoryProvider#MyFactoryProvider$superArg$1(this.mpep, this.locator), 
    [UNKNOWN : org.glassfish.jersey.model.Parameter.Source]
  ) {
    private val mpep: 
      
        org.glassfish.jersey.server.internal.inject.
          MultivaluedParameterExtractorProvider
      
    private val locator: org.glassfish.hk2.api.ServiceLocator
    override def createValueProvider(
      parameter: org.glassfish.jersey.server.model.Parameter
    ): 
      
        java.util.function.Function[org.glassfish.jersey.server.ContainerRequest
          , 
        ?]
      
     = 
      {
        def $anonfun(request: org.glassfish.jersey.server.ContainerRequest): Any
           = 
        ""
        closure($anonfun:
          
            java.util.function.Function[
              org.glassfish.jersey.server.ContainerRequest
            , ?]
          
        )
      }
    private <static> def MyFactoryProvider$superArg$1(
      mpep: 
        
          org.glassfish.jersey.server.internal.inject.
            MultivaluedParameterExtractorProvider
        
    , locator: org.glassfish.hk2.api.ServiceLocator): 
      
        javax.inject.Provider[
          org.glassfish.jersey.server.internal.inject.
            MultivaluedParameterExtractorProvider
        ]
      
     = 
      {
        final class $anon() extends Object(), 
          javax.inject.Provider[
            org.glassfish.jersey.server.internal.inject.
              MultivaluedParameterExtractorProvider
          ]
         {
          override def get(): 
            
              org.glassfish.jersey.server.internal.inject.
                MultivaluedParameterExtractorProvider
            
           = MyFactoryProvider.this.mpep
        }
        new Object with javax.inject.Provider {...}():
          
            javax.inject.Provider[
              org.glassfish.jersey.server.internal.inject.
                MultivaluedParameterExtractorProvider
            ]
          
      }
  }
} of class class dotty.tools.dotc.ast.Trees$PackageDef # 28132

java.lang.AssertionError: assertion failed: asTerm called on not-a-Term val <none> while compiling /Users/jens/Developer/spring/src/main/scala/Main.scala
[error] ## Exception when compiling 1 sources to /Users/jens/Developer/spring/target/scala-0.24/classes
[error] java.lang.AssertionError: assertion failed: asTerm called on not-a-Term val <none>
[error] dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
[error] dotty.tools.dotc.core.Symbols$Symbol.asTerm(Symbols.scala:532)
[error] dotty.tools.dotc.transform.ExplicitOuter$.dotty$tools$dotc$transform$ExplicitOuter$$$outerParamAccessor(ExplicitOuter.scala:222)
[error] dotty.tools.dotc.transform.ExplicitOuter$OuterOps$.loop$1(ExplicitOuter.scala:405)
[error] dotty.tools.dotc.transform.ExplicitOuter$OuterOps$.path$extension(ExplicitOuter.scala:413)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedThis(Erasure.scala:721)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2336)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2394)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:124)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2433)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2442)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2445)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedSelect(Erasure.scala:640)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2315)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2393)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:124)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2433)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2442)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2445)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2556)
[error] dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:1867)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:866)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2322)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2393)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:124)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2433)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2442)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2445)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2467)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2512)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:953)
[error] dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2003)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2325)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2393)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:124)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2433)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2442)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2445)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2467)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2512)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:953)
[error] dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:849)
[error] dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:853)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2343)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2394)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:124)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2433)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2442)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2445)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2556)
[error] dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:1867)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:866)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2322)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2393)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:124)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2433)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2442)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2445)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2467)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2512)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:953)
[error] dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2003)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2325)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2393)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:124)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2433)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2442)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2445)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2467)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2512)
[error] dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:953)
[error] dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2129)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2366)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2394)
[error] dotty.tools.dotc.typer.ReTyper.typedUnadapted(ReTyper.scala:124)
[error] dotty.tools.dotc.typer.Typer.op$1(Typer.scala:2433)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2442)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2445)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:2556)
[error] dotty.tools.dotc.transform.Erasure.run(Erasure.scala:121)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:318)
[error] scala.collection.immutable.List.map(List.scala:219)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:319)
[error] dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:165)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$5(Run.scala:175)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:183)
[error] dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:64)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:190)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:127)
[error] dotty.tools.dotc.Run.compile(Run.scala:110)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:38)
[error] dotty.tools.dotc.Driver.process(Driver.scala:194)
[error] dotty.tools.dotc.Main.process(Main.scala)
[error] xsbt.CachedCompilerImpl.run(CachedCompilerImpl.java:69)
[error] xsbt.CompilerInterface.run(CompilerInterface.java:41)
[error] sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] java.lang.reflect.Method.invoke(Method.java:498)
[error] sbt.internal.inc.AnalyzingCompiler.call(AnalyzingCompiler.scala:248)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:122)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:95)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:91)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:186)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$3(MixedAnalyzingCompiler.scala:82)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$3$adapted(MixedAnalyzingCompiler.scala:77)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:215)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:77)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:146)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:343)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:343)
[error] sbt.internal.inc.Incremental$.doCompile(Incremental.scala:120)
[error] sbt.internal.inc.Incremental$.$anonfun$compile$4(Incremental.scala:100)
[error] sbt.internal.inc.IncrementalCommon.recompileClasses(IncrementalCommon.scala:180)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:98)
[error] sbt.internal.inc.Incremental$.$anonfun$compile$3(Incremental.scala:102)
[error] sbt.internal.inc.Incremental$.manageClassfiles(Incremental.scala:155)
[error] sbt.internal.inc.Incremental$.compile(Incremental.scala:92)
[error] sbt.internal.inc.IncrementalCompile$.apply(Compile.scala:75)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:348)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:301)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:168)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:248)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:74)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:1762)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:1735)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:67)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:281)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:19)
[error] sbt.Execute.work(Execute.scala:290)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:281)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:178)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:37)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] java.lang.Thread.run(Thread.java:748)
[error]            
[error] stack trace is suppressed; run last Compile / compileIncremental for the full output
[error] (Compile / compileIncremental) java.lang.AssertionError: assertion failed: asTerm called on not-a-Term val <none>
```
</details>

--- a/compiler/src/dotty/tools/dotc/ast/TreeTypeMap.scala
+++ b/compiler/src/dotty/tools/dotc/ast/TreeTypeMap.scala
@@ -91,7 +91,7 @@ class TreeTypeMap(
         case id: Ident if tpd.needsSelect(id.tpe) =>
           ref(id.tpe.asInstanceOf[TermRef]).withSpan(id.span)
         case ddef @ DefDef(name, tparams, vparamss, tpt, _) =>
-          val (tmap1, tparams1) = transformDefs(ddef.tparams)
+          val (tmap1, tparams1) = transformDefs(tparams)
           val (tmap2, vparamss1) = tmap1.transformVParamss(vparamss)
           val res = cpy.DefDef(ddef)(name, tparams1, vparamss1, tmap2.transform(tpt), tmap2.transform(ddef.rhs))
           res.symbol.setParamssFromDefs(tparams1, vparamss1)

--- a/compiler/src/dotty/tools/dotc/transform/HoistSuperArgs.scala
+++ b/compiler/src/dotty/tools/dotc/transform/HoistSuperArgs.scala
@@ -33,7 +33,7 @@ object HoistSuperArgs {
  *
  *  An argument is complex if it contains a method or template definition, a this or a new,
  *  or it contains an identifier which needs a `this` prefix to be accessed. This is the case
- *  if the identifer neither a global reference nor a reference to a parameter of the enclosing class.
+ *  if the identifier has neither a global reference nor a reference to a parameter of the enclosing class.
  *  @see needsHoist for an implementation.
  *
  *  A hoisted argument definition gets the parameters of the class it is hoisted from
@@ -127,7 +127,7 @@ class HoistSuperArgs extends MiniPhase with IdentityDenotTransformer { thisPhase
       arg match {
         case Apply(fn, arg1 :: Nil) if fn.symbol == defn.cbnArg =>
           cpy.Apply(arg)(fn, hoistSuperArg(arg1, cdef) :: Nil)
-        case _ if (arg.existsSubTree(needsHoist)) =>
+        case _ if arg.existsSubTree(needsHoist) =>
           val superMeth = newSuperArgMethod(arg.tpe)
           val superArgDef = polyDefDef(superMeth, trefs => vrefss => {
             val paramSyms = trefs.map(_.typeSymbol) ::: vrefss.flatten.map(_.symbol)

--- a/compiler/src/dotty/tools/dotc/transform/HoistSuperArgs.scala
+++ b/compiler/src/dotty/tools/dotc/transform/HoistSuperArgs.scala
@@ -116,6 +116,13 @@ class HoistSuperArgs extends MiniPhase with IdentityDenotTransformer { thisPhase
         case _                    => false
       }
 
+      /** Only rewire types that are owned by the current Hoister and is an param or accessor */
+      def needsRewire(tp: Type) = tp match {
+        case ntp: NamedType =>
+          (ntp.symbol.owner == cls || ntp.symbol.owner == constr) && ntp.symbol.isParamOrAccessor
+        case _ => false
+      }
+
       // begin hoistSuperArg
       arg match {
         case Apply(fn, arg1 :: Nil) if fn.symbol == defn.cbnArg =>
@@ -128,9 +135,7 @@ class HoistSuperArgs extends MiniPhase with IdentityDenotTransformer { thisPhase
               typeMap = new TypeMap {
                 lazy val origToParam = origParams.zip(paramSyms).toMap
                 def apply(tp: Type) = tp match {
-                  case tp: NamedType
-                  if (tp.symbol.owner == cls || tp.symbol.owner == constr) &&
-                     tp.symbol.isParamOrAccessor =>
+                  case tp: NamedType if needsRewire(tp) =>
                     origToParam.get(tp.symbol) match {
                       case Some(mappedSym) => if (tp.symbol.isType) mappedSym.typeRef else mappedSym.termRef
                       case None => mapOver(tp)
@@ -140,7 +145,7 @@ class HoistSuperArgs extends MiniPhase with IdentityDenotTransformer { thisPhase
                 }
               },
               treeMap = {
-                case tree: RefTree if paramSyms.contains(tree.symbol) =>
+                case tree: RefTree if needsRewire(tree.tpe) =>
                   cpy.Ident(tree)(tree.name).withType(tree.tpe)
                 case tree =>
                   tree

--- a/compiler/src/dotty/tools/dotc/transform/ProtectedAccessors.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ProtectedAccessors.scala
@@ -10,7 +10,7 @@ import MegaPhase.MiniPhase
 import config.Printers.transforms
 
 /** Add accessors for all protected accesses. An accessor is needed if
- *  according to the rules of the JVM a protected class member is not accesissible
+ *  according to the rules of the JVM a protected class member is not accessible
  *  from the point of access, but is accessible if the access is from an enclosing
  *  class. In this point a public access method is placed in that enclosing class.
  */

--- a/tests/pos/i8748.scala
+++ b/tests/pos/i8748.scala
@@ -1,0 +1,9 @@
+class A(a: String) extends B(new C {
+  override def get(): String = a
+})
+
+class B(c: C)
+
+trait C {
+  def get(): String
+}

--- a/tests/pos/i8786.scala
+++ b/tests/pos/i8786.scala
@@ -1,0 +1,7 @@
+class B(y: Int) extends A(new C(y){})
+
+class A(c: C)
+
+abstract class C(y: Int) {
+  def x: Int = y
+}

